feat: Add an option to the HTTP In to include the raw body.

debadutta98 opened the pull request:
<!--
## Before you hit that Submit button....

Please read our [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
before submitting a pull request.

## Types of changes

What types of changes does your code introduce?
Put an `x` in the boxes that apply
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

<!--
If you want to raise a pull-request with a new feature, or a refactoring
of existing code, it **may well get rejected** if it hasn't been discussed on
the [forum](https://discourse.nodered.org) or
[slack team](https://nodered.org/slack) first.

-->

## Proposed changes

This PR Closes: #5029

## How is it implemented?

This PR allows users to access the raw request body by accessing it from the request object. It achieves this by programmatically adding middleware at the top of the app's route stack.

When a request is received, the middleware is triggered and attaches event listeners to the request stream. These listeners clone the stream, preserving the original data so it can be consumed later and converted into raw data.

## What happens after the request passes through middleware?

When a new HTTP In node is initialized, I generate its unique route key and store it in a hash set. Later, when an incoming request passes through the middleware, I check whether the route key exists in the hash set. If the key does not exist, there's no need to clone the request stream. However, if the key is found, I clone the stream to preserve the raw request data.

![Screenshot from 2025-02-03 15-07-28](https://github.com/user-attachments/assets/6401debc-5c21-4c2b-b820-1229617e5248)

## Checklist
<!-- Put an `x` in the boxes that apply -->

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [ ] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `npm run test` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

debadutta98 commented:
Hi @knolleary, This PR is now ready for review after fixing the following issues:

- Resolved the Node-RED embedding issue by programmatically injecting the middleware at the top of the middleware stack.
- Fixed a timeout issue that occurred when changing the node's position by correcting the method.

debadutta98 commented:
Hi @knolleary , can you please review this PR and give your valuable feedback so that I would know needs to be improved.

Thank

knolleary commented:
Can you rebase your branch on the latest `dev` as this PR appears to be pulling in other commits.

I don't have time this week or next (due to holiday) to review this properly. It's a more complicated approach than I was thinking of and need to spend some time understanding how the stream logic works and what side-effects it could have.

I'm sorry that I can't give you a quicker response than that.

debadutta98 commented:
Hi @knolleary,

Thanks for reviewing this PR.

I have rebased this branch with the dev branch.

To clarify my earlier comment, this PR ensures that data is held in a stream until it is consumed. Since Node.js readable streams can only be consumed once, we create a clone of the stream by adding an event listener to the request stream. This listener pushes the data to the cloned stream, which holds the data until it is consumed by the raw data middleware. The cloned stream is set in the request context and is only accessed within the middleware.

Note: The stream is cloned only when the user enables raw data in the UI, preventing unnecessary memory usage.

Looking forward to your further feedback and review.

Thank you!

debadutta98 commented:
Hi @knolleary,

Hope you are doing well

Just a quick reminder, could you please review this PR when you get a chance? Let me know what can be improved so we can collaborate on it.

knolleary commented on packages/node_modules/@node-red/nodes/core/network/21-httpin.js:
```diff
@@ -26,16 +27,21 @@ module.exports = function(RED) {
+    var PassThrough = require('stream').PassThrough;
+    var rawDataRoutes = new Set();
 
-    function rawBodyParser(req, res, next) {
-        if (req.skipRawBodyParser) { next(); } // don't parse this if told to skip
```
This is removing important logic we need to have.

1. `req.skipRawBodyParser` is a valid flag users can set in custom middleware to avoid any parsing of the body. I can't remember the full reason we have that flag, but we do - so it must be honoured.
2. Your logic will bypass the raw body parser if `_nodeRedReqStream` isn't set. However that is only set if its a route where the user has explicitly asked for the raw body to be available in *addition* to any parsed version of the payload. If the request doesn't provide a content-type that can be parsed before this middleware, we need this middleware to get the raw body.

My main concern with this approach is that we end up with two copies of the payload in memory.

An alternative approach would be to add an option to 'skip body parsing', which will bypass any content-type based parsing and just provide the payload as a Buffer object. The user is then free to handle that as they need -  using the JSON node to parse if that's needed downstream.

debadutta98 replied:
1. Yes, I agree with this point. As a resolution, I will retain all the variables used in the code without removing or altering their usage.
2. I also agree that the current approach could be quite expensive from a memory utilization perspective. The alternative approach you suggested is more memory-efficient, so I have implemented it as recommended.

knolleary reviewed (changes requested):
Some comments on your PR - apologies for the delay.

The more I think about this feature, I'm thinking a better alternative approach would be to add an option to 'skip body parsing', which will bypass any content-type based parsing and just provide the payload as a Buffer object. This will avoid having two copies of the body in memory. The user is then free to handle that as they need - using the JSON node to parse if that's needed downstream.

What do you think?

I also noticed your message catalogue updates need double checking - you are adding the translated message in different places in each language.

debadutta98 commented:
Hi @knolleary ,

Thank you for reviewing and considering this PR.

As you highlighted, the current approach is not memory-efficient since it stores duplicate copies of the data in memory. This can lead to memory leaks if the data size exceeds the available heap memory. Your suggestion to skip body parsing is a much better alternative for optimizing memory usage. By providing an option for users to skip body parsing, we can store the raw data as a buffer, allowing users to process it further as needed.

I have implemented all the suggested changes, and everything looks good to me. Additionally, I removed all language translations that I’m not familiar with to prevent any potential mistranslation.

![Screenshot from 2025-03-18 19-11-42](https://github.com/user-attachments/assets/1ad1ded7-05d9-4d96-b1d1-dfd535d2b1ea)

Waiting for your feedback 

Thank you!

debadutta98 commented:
Hi @knolleary , could you please review this PR and let me know what could be improve or done on this PR.

Thank you

knolleary commented:
@debadutta98 thank you for your patience. I've just tested your PR and it looks good. I may tweak the UI label slightly to be a bit clearer to the end user; but basically I'm happy for this to merge as-is and I'll make a couple minor changes along the way. This will be in Node-RED 4.1 - thank you!
